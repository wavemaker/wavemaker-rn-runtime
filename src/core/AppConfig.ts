export interface Drawer {
    setContent: (c: React.ReactNode) => void;
    getContent: () => React.ReactNode;
    setAnimation: (animation: string) => void;
    getAnimation: () => string;
}

export default interface AppConfig {
    assets: any;
    appProperties: any;
    appLocale: any;
    url: string;
    loadApp: boolean;
    refresh: (complete?: boolean) => void,
    currentPage?: any;
    pages?: any[];
    landingPage: string;
    partials?: any[];
    drawer: Drawer;
    app: any;
    setDrawerContent: any;
    theme: any;
    drawerType: any;
    getServiceDefinitions: any;
    loggedInUser: any;
    selectedLocale: string;
}