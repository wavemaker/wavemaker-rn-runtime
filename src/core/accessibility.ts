import { AccessibilityInfo } from 'react-native';
import { isAndroid, removeUndefinedKeys } from './utils';

let _isScreenReaderEnabled = false;

AccessibilityInfo.addEventListener(
  'screenReaderChanged',
  flag => {
    _isScreenReaderEnabled = flag;
  },
);

export const isScreenReaderEnabled = () => _isScreenReaderEnabled;

async function getScreenReaderStatus() {
  _isScreenReaderEnabled = await AccessibilityInfo.isScreenReaderEnabled();
}

getScreenReaderStatus();

export enum AccessibilityWidgetType {
  BUTTON = 'button',
  PICTURE = 'picture',
  TEXT = 'text',
  NUMBER = 'number',
  TEXTAREA = 'textarea',
  SELECT = 'select',
  CHIPS = 'chips',
  CURRENCY = 'currency',
  RADIOSET = 'radioset',
  CHECKBOX = 'checkbox',
  TOGGLE = 'toggle',
  SWITCH = 'switch',
  DATE = 'date',
  VIDEO = 'video',
  PROGRESSBAR = 'progressbar',
  PROGRESSCIRCLE = 'progresscircle',
  LABEL = 'label',
  ANCHOR = 'anchor',
  MESSAGE = 'message',
  SEARCH = 'search',
  ICON = 'icon',
  NAV = 'nav',
  POVOVER = 'popover',
  WEBVIEW = 'webview',
  LINECHART = 'linechart',
};

  
export type AccessibilityPropsType = {
  accessible?: boolean;
  accessibilityLabel?: string;
  accessibilityLabelledBy?: string;
  accessibilityHint?: string;
  accessibilityRole?: 'button' | 'link' | 'header' | 'search' | 'image' | 'imagebutton' | 'none' | 'summary' | 'text' | 'progressbar' | 'grid' | 'alert';
  accessibilityState?: {
    disabled?: boolean;
    selected?: boolean;
    checked?: boolean;
    expanded?: boolean;
  };
  accessibilityValue?: {
    min?: number;
    max?: number;
    now?: number;
    text?: string;
  };
  accessibilityActions?: Array<{
    name: string;
    label?: string;
  }>;
  accessibilityLiveRegion?: 'none' | 'polite' | 'assertive';
  accessibilityLanguage?: any;
  accessibilityElementsHidden?: boolean;
  accessibilityViewIsModal?: boolean;
};
  

export const getAccessibilityProps = (widgetType: AccessibilityWidgetType, accessibilityProps: AccessibilityPropsType | any) => {
  let props: AccessibilityPropsType = {accessible: true};
  if (!_isScreenReaderEnabled) {
    return {};
  }
  switch (widgetType) {
    case AccessibilityWidgetType.BUTTON:
    case AccessibilityWidgetType.TEXT:
    case AccessibilityWidgetType.NUMBER:
    case AccessibilityWidgetType.TEXTAREA:
    case AccessibilityWidgetType.SELECT:
    case AccessibilityWidgetType.CURRENCY:
    case AccessibilityWidgetType.TOGGLE:
    case AccessibilityWidgetType.DATE:
    case AccessibilityWidgetType.LABEL:
    case AccessibilityWidgetType.ANCHOR:
    case AccessibilityWidgetType.MESSAGE:    
    case AccessibilityWidgetType.SEARCH: 
    case AccessibilityWidgetType.PICTURE: 
    case AccessibilityWidgetType.ICON:
    case AccessibilityWidgetType.NAV:
    case AccessibilityWidgetType.POVOVER:
    case AccessibilityWidgetType.WEBVIEW:
    case AccessibilityWidgetType.LINECHART:  
    case AccessibilityWidgetType.VIDEO: {
      props.accessibilityLabel = accessibilityProps.accessibilitylabel || accessibilityProps.caption;
      props.accessibilityHint = accessibilityProps.hint;
      props.accessibilityRole = accessibilityProps.accessibilityrole;

      if (
        widgetType === AccessibilityWidgetType.BUTTON ||
        widgetType === AccessibilityWidgetType.TEXT ||
        widgetType === AccessibilityWidgetType.NUMBER ||
        widgetType === AccessibilityWidgetType.TEXTAREA ||
        widgetType === AccessibilityWidgetType.SELECT ||
        widgetType === AccessibilityWidgetType.TOGGLE ||
        widgetType === AccessibilityWidgetType.DATE
      ) {
        props.accessibilityState = { disabled: accessibilityProps.disabled };
      }
      if (
        (widgetType === AccessibilityWidgetType.TEXT ||
          widgetType === AccessibilityWidgetType.NUMBER ||
          widgetType === AccessibilityWidgetType.TEXTAREA ||
          widgetType === AccessibilityWidgetType.SELECT ||
          widgetType === AccessibilityWidgetType.CURRENCY ||
          widgetType === AccessibilityWidgetType.TOGGLE) &&
        isAndroid()
      ) {
        props.accessibilityLabelledBy =
          accessibilityProps.accessibilitylabelledby;
      }
      if (
        widgetType === AccessibilityWidgetType.NUMBER ||
        widgetType === AccessibilityWidgetType.CURRENCY
      ) {
        props.accessibilityValue = {
          min: accessibilityProps.minvalue,
          max: accessibilityProps.maxvalue,
        };
      }
      if (widgetType === AccessibilityWidgetType.SELECT) {
        props.accessibilityState = {
          ...props.accessibilityState,
          expanded: accessibilityProps.expanded,
        };
      }
      if (widgetType === AccessibilityWidgetType.TOGGLE) {
        props.accessibilityState = {
          ...props.accessibilityState,
          selected: accessibilityProps.selected,
        };
      }
      break;
    }

    case AccessibilityWidgetType.CHIPS: {
      props.accessibilityLabel = accessibilityProps.accessibilitylabel || accessibilityProps.caption;
      props.accessibilityHint = accessibilityProps.hint;
      props.accessibilityState = {
        disabled: accessibilityProps.disabled,
        selected: accessibilityProps.selected,
      };
      break;
    }

    case AccessibilityWidgetType.RADIOSET: {
      props.accessibilityState = {
        disabled: accessibilityProps.readonly || accessibilityProps.disabled,
        selected: accessibilityProps.selected,
      };
      break;
    }

    case AccessibilityWidgetType.CHECKBOX: {
      props.accessibilityState = {
        disabled: accessibilityProps.readonly ||  accessibilityProps.disabled,
        checked: accessibilityProps.checked,
      };
      break;
    }

    case AccessibilityWidgetType.SWITCH: {
      props.accessibilityState = {
        disabled: accessibilityProps.disabled,
        selected: accessibilityProps.selected,
      };
      break;
    }
    case AccessibilityWidgetType.PROGRESSBAR:
    case AccessibilityWidgetType.PROGRESSCIRCLE: {
      props.accessibilityLabel = accessibilityProps.accessibilitylabel || accessibilityProps.caption;
      props.accessibilityRole = accessibilityProps.accessibilityrole;
      break;
    }
    default:
      break;
  }

  const finalProps = removeUndefinedKeys(props);
  // console.log('finalProps', finalProps)

  return finalProps;
}
  