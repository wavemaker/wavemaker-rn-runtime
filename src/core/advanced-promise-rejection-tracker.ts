/**
 * Advanced Promise Rejection Tracker
 *
 * Provides comprehensive promise rejection tracking across different JavaScript engines.
 * This module addresses the critical gap where API call failures and async function errors
 * are not caught by React Error Boundaries, ensuring all promise rejections are handled.
 *
 * Engine Support:
 * - Web: Uses browser's 'unhandledrejection' event
 * - Hermes: Uses HermesInternal.enablePromiseRejectionTracker
 * - JSC: Uses React Native's 'promise/setimmediate/rejection-tracking' module
 *
 * Coverage Includes:
 * - API call failures (fetch, axios, any HTTP library)
 * - Async function errors (unhandled throws in async/await)
 * - Manual Promise.reject() calls
 * - Third-party library promise rejections
 *
 * Research Notes:
 * This implementation was developed through extensive research into React Native's
 * promise handling across different engines. Each engine has its own optimal approach:
 * - Hermes provides a native tracking API that hooks into the engine's internals
 * - JSC can use React Native's internal promise tracking module
 * - Web browsers have standardized unhandledrejection events
 *
 * The approach automatically detects the best available method and falls back
 * gracefully when native tracking is unavailable.
 */

import { Platform } from 'react-native';

// TypeScript declarations for global objects
declare global {
  interface Global {
    HermesInternal?: {
      enablePromiseRejectionTracker?: (options: {
        allRejections: boolean;
        onUnhandled: (id: number, error: any) => void;
        onHandled: (id: number) => void;
      }) => void;
    };
  }
}

export interface PromiseRejectionHandler {
  (error: any): void;
}

/**
 * Multi-engine promise rejection tracking implementation
 */
export class AdvancedPromiseRejectionTracker {
  private onRejectionHandler: PromiseRejectionHandler;
  private cleanupFunctions: (() => void)[] = [];

  constructor(onRejection: PromiseRejectionHandler) {
    this.onRejectionHandler = onRejection;
  }

  /**
   * Setup promise rejection tracking based on available APIs
   * Returns true if native tracking was enabled, false if fallback used
   */
  public setup(): boolean {
    // Automatically detect and setup the best available promise rejection tracking method

    if (Platform.OS === 'web') {
      return this.setupWebTracking();
    }

    // For mobile platforms, try native tracking methods in order of preference
    if (this.setupHermesTracking()) {
      return true;
    }

    // Fallback to React Native internal tracking for JSC
    if (this.setupReactNativeTracking()) {
      return true;
    }

    // No native tracking available - this should rarely happen in modern RN versions
    return false;
  }

  /**
   * Method 1: Web Environment - Use unhandledrejection event
   */
  private setupWebTracking(): boolean {
    try {
      if (typeof window === 'undefined' || !window.addEventListener) {
        return false;
      }

      const handler = (event: PromiseRejectionEvent) => {
        event.preventDefault(); // Prevent browser's default handling
        const error =
          event.reason instanceof Error
            ? event.reason
            : new Error(String(event.reason));
        this.onRejectionHandler(error);
      };

      window.addEventListener('unhandledrejection', handler);
      this.cleanupFunctions.push(() => {
        window.removeEventListener('unhandledrejection', handler);
      });

      return true;
    } catch (error) {
      return false;
    }
  }

  /**
   * Method 2: Hermes Engine - Use HermesInternal API
   */
  private setupHermesTracking(): boolean {
    try {
      // Check if Hermes engine is available
      const hermesInternal = (global as any).HermesInternal;
      if (
        hermesInternal &&
        typeof hermesInternal.enablePromiseRejectionTracker === 'function'
      ) {
        hermesInternal.enablePromiseRejectionTracker({
          allRejections: true, // Track both handled and unhandled rejections
          onUnhandled: (id: number, error: any) => {
            this.onRejectionHandler(error);
          },
          onHandled: (id: number) => {
            // Promise was handled later - could be used for analytics in the future
          },
        });

        // Cleanup function (if available)
        this.cleanupFunctions.push(() => {
          // Hermes doesn't provide a disable method, but we can set a no-op handler
          if (hermesInternal?.enablePromiseRejectionTracker) {
            hermesInternal.enablePromiseRejectionTracker({
              allRejections: false,
              onUnhandled: () => {},
              onHandled: () => {},
            });
          }
        });

        return true;
      }
      return false;
    } catch (error) {
      return false;
    }
  }

  /**
   * Method 3: React Native Internal - Use promise/setimmediate/rejection-tracking for JSC
   */
  private setupReactNativeTracking(): boolean {
    try {
      // Try to access React Native's internal promise tracking module
      const tracking = require('promise/setimmediate/rejection-tracking');

      if (tracking && typeof tracking.enable === 'function') {
        tracking.enable({
          allRejections: true,
          onUnhandled: (id: number, error: any) => {
            this.onRejectionHandler(error);
          },
          onHandled: (id: number) => {
            // Promise was handled later - tracking for potential analytics
          },
        });

        this.cleanupFunctions.push(() => {
          // Try to disable tracking if possible
          if (tracking.disable && typeof tracking.disable === 'function') {
            tracking.disable();
          }
        });

        return true;
      }
      return false;
    } catch (error) {
      return false;
    }
  }

  /**
   * Clean up all tracking mechanisms
   */
  public cleanup(): void {
    // Clean up all tracking mechanisms and restore original state
    this.cleanupFunctions.forEach((cleanup) => {
      try {
        cleanup();
      } catch (error) {
        // Silently handle cleanup errors to avoid disrupting app shutdown
      }
    });
    this.cleanupFunctions = [];
  }
}

/**
 * Helper function to detect JavaScript engine (simplified for Web/Hermes/JSC)
 */
export function detectJavaScriptEngine(): 'web' | 'hermes' | 'jsc' {
  if (Platform.OS === 'web') {
    return 'web';
  }

  if (typeof (global as any).HermesInternal !== 'undefined') {
    return 'hermes';
  }

  // For non-Hermes engines on mobile, assume JSC (most common)
  return 'jsc';
}

/**
 * Public API for easy integration
 */
/**
 * Sets up advanced promise rejection tracking with automatic engine detection
 * @param onRejection Callback function to handle unhandled promise rejections
 * @returns AdvancedPromiseRejectionTracker instance for cleanup and testing
 */
export function setupAdvancedPromiseRejectionTracking(
  onRejection: PromiseRejectionHandler
): AdvancedPromiseRejectionTracker {
  const tracker = new AdvancedPromiseRejectionTracker(onRejection);
  tracker.setup();
  return tracker;
}
