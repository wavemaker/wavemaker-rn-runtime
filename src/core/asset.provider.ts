import React from 'react';

const AssetContext = React.createContext<(path: string) => number | string>(null as any);

export const AssetProvider = AssetContext.Provider;
export const AssetConsumer = AssetContext.Consumer;