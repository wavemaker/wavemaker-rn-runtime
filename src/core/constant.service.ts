export class ConstantService {
   VALIDATOR = {
    REQUIRED: 'required',
    MAXCHARS: 'maxchars',
    MINVALUE: 'minvalue',
    MAXVALUE: 'maxvalue',
    REGEXP: 'regexp',
    MINDATE: 'mindate',
    MAXDATE: 'maxdate',
    MINTIME: 'mintime',
    MAXTIME: 'maxtime',
    EXCLUDEDATES: 'excludedates',
    EXCLUDEDAYS: 'excludedays'
  };
}

export default new ConstantService();
