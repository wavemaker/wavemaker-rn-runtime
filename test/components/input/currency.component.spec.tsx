import React, { createRef } from 'react';
import {
  act,
  cleanup,
  fireEvent,
  render,
  waitFor,
} from '@testing-library/react-native';
import WmCurrency from '@wavemaker/app-rn-runtime/components/input/currency/currency.component';
import { CURRENCY_INFO } from '@wavemaker/app-rn-runtime/core/currency-constants';
import { Platform, TextInput } from 'react-native';

const defaultProps = {
  currency: 'USD',
  placeholder: 'Enter amount',
  datavalue: null,
  minvalue: null,
  maxvalue: null,
  step: 1,
  required: false,
  readonly: false,
  disabled: false,
  decimalPlaces: 2,
};

describe('Test Currency component', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
    jest.clearAllMocks();
    cleanup();
  });

  test('should render correctly with default props', () => {
    const tree = render(
      <WmCurrency {...defaultProps} floatinglabel="Amount" />
    );
    const { getByText, UNSAFE_getByType } = tree;

    expect(getByText(CURRENCY_INFO[defaultProps.currency].symbol)).toBeTruthy();
    expect(UNSAFE_getByType(TextInput)).toBeTruthy();
    expect(getByText('Amount')).toBeTruthy();
    expect(tree).toMatchSnapshot();
  });

  test('should set the currency symbol on props.currency change', async () => {
    const { getByText, rerender } = render(<WmCurrency {...defaultProps} />);
    expect(getByText(CURRENCY_INFO[defaultProps.currency].symbol)).toBeTruthy();

    rerender(<WmCurrency {...defaultProps} currency="EUR" />);
    // Ensure the new currency symbol is present
    await waitFor(() => {
      expect(getByText(CURRENCY_INFO['EUR'].symbol)).toBeTruthy();
    });
  });

  test('should handle input change correctly', () => {
    const onChangeTextMock = jest.spyOn(WmCurrency.prototype, 'onChangeText');
    const { getByPlaceholderText } = render(<WmCurrency {...defaultProps} />);
    const input = getByPlaceholderText(defaultProps.placeholder);

    fireEvent.changeText(input, '123.45');
    expect(onChangeTextMock).toHaveBeenCalledWith('123.45', 'currency');
  });

  test('should validate number correctly', () => {
    const onChangeTextMock = jest.spyOn(WmCurrency.prototype, 'onChangeText');
    const { getByPlaceholderText } = render(<WmCurrency {...defaultProps} />);
    const input = getByPlaceholderText(defaultProps.placeholder);

    fireEvent.changeText(input, '123.45');
    fireEvent(input, 'blur', {
      target: {
        value: null,
      },
    });

    // Check if the value is set correctly
    expect(onChangeTextMock).toHaveBeenCalledWith('123.45', 'currency');

    // Check invalid values
    fireEvent.changeText(input, '12.34563');
    fireEvent(input, 'blur', {
      target: {
        value: null,
      },
    });

    // Should remain unchanged if invalid
    expect(onChangeTextMock).toHaveBeenCalledWith('123.45', 'currency');
  });

  xit('should not accept negative number', async () => {
    const updateStateMock = jest.spyOn(WmCurrency.prototype, 'updateState');
    const onChangeTextMock = jest.spyOn(WmCurrency.prototype, 'onChangeText');
    const customRef = createRef<WmCurrency>();
    const { getByPlaceholderText } = render(
      <WmCurrency {...defaultProps} ref={customRef} />
    );
    const input = getByPlaceholderText(defaultProps.placeholder);

    fireEvent.changeText(input, '-123.45');
    fireEvent(input, 'blur', {
      target: {
        value: null,
      },
    });

    // Check if the value is set correctly
    expect(onChangeTextMock).toHaveBeenCalledWith('-123.45', 'currency');
    await waitFor(() => {
      expect(customRef.current.state.textValue).toBe(null);
    });
  });

  test('should validate number correctly, only supports "e" as a character', () => {
    const onChangeTextMock = jest.spyOn(WmCurrency.prototype, 'onChangeText');
    const { getByPlaceholderText } = render(<WmCurrency {...defaultProps} />);
    const input = getByPlaceholderText(defaultProps.placeholder);

    fireEvent.changeText(input, '123.45');
    fireEvent(input, 'blur', {
      target: {
        value: null,
      },
    });

    expect(onChangeTextMock).toHaveBeenCalledWith('123.45', 'currency');

    fireEvent.changeText(input, '10e');
    fireEvent(input, 'blur', {
      target: {
        value: null,
      },
    });

    expect(onChangeTextMock).toHaveBeenCalledWith('10e', 'currency');

    fireEvent.changeText(input, '10a');
    fireEvent(input, 'blur', {
      target: {
        value: null,
      },
    });

    expect(onChangeTextMock).toHaveBeenCalledWith('10e', 'currency');

    fireEvent.changeText(input, '$$');
    fireEvent(input, 'blur', {
      target: {
        value: null,
      },
    });

    expect(onChangeTextMock).toHaveBeenCalledWith('10e', 'currency');
  });

  test('should enable input if disabled and readonly props are falsy', () => {
    const { getByPlaceholderText, rerender } = render(
      <WmCurrency {...defaultProps} readonly={true} disabled={true} />
    );
    const input = getByPlaceholderText(defaultProps.placeholder);
    expect(input.props.editable).toBe(false);

    rerender(
      <WmCurrency {...defaultProps} readonly={false} disabled={false} />
    );
    expect(input.props.editable).toBe(true);
  });

  test('should disable input if disabled is true', () => {
    const { getByPlaceholderText, rerender } = render(
      <WmCurrency {...defaultProps} />
    );
    const input = getByPlaceholderText(defaultProps.placeholder);

    rerender(<WmCurrency {...defaultProps} disabled={true} />);
    expect(input.props.editable).toBe(false);
  });

  test('should disable input if readonly props is true', () => {
    const { getByPlaceholderText, rerender } = render(
      <WmCurrency {...defaultProps} />
    );
    const input = getByPlaceholderText(defaultProps.placeholder);

    rerender(<WmCurrency {...defaultProps} readonly={true} />);
    expect(input.props.editable).toBe(false);
  });

  xit('should validate against required prop correctly', () => {
    const customRef = createRef();
    const { getByPlaceholderText } = render(
      <WmCurrency {...defaultProps} ref={customRef} updateon="default" />
    );
    const input = getByPlaceholderText(defaultProps.placeholder);

    fireEvent.changeText(input, '');
    fireEvent(input, 'blur', {
      target: {
        value: null,
      },
    });

    // Check if input indicates invalid state
    expect(customRef.current.state.isValid).toBe(false);
  });

  test('should have default value for native platform', () => {
    const tree = render(
      <WmCurrency {...defaultProps} datavalue="sample text" />
    );
    expect(tree.UNSAFE_getByType(TextInput).props.defaultValue).toBe(
      'sample text'
    );
    expect(tree).toMatchSnapshot();
  });

  test('should have default value for "platform=web"', () => {
    (Platform as any).OS = 'web';

    const tree = render(
      <WmCurrency {...defaultProps} datavalue="sample text" />
    );
    expect(tree.UNSAFE_getByType(TextInput).props.value).toBe('sample text');
    expect(tree).toMatchSnapshot();
  });

  test('should not show component when show prop is false', () => {
    const tree = render(<WmCurrency {...defaultProps} show={false} />);

    expect(tree.toJSON().props.style).toMatchObject({
      height: 0,
      width: 0,
    });
  });

  xit('should isValid false when input number is below minvalue', async () => {
    const onChangeTextMock = jest.spyOn(WmCurrency.prototype, 'onChangeText');
    const customRef = createRef();
    const tree = render(
      <WmCurrency
        {...defaultProps}
        minvalue={100}
        // maxvalue={1000}
        ref={customRef}
        updateon="default"
      />
    );
    const input = tree.getByPlaceholderText('Enter amount');

    fireEvent.changeText(input, '500');
    fireEvent(input, 'blur', {
      target: {
        value: null,
      },
    });

    await waitFor(() => {
      expect(onChangeTextMock).toHaveBeenCalledWith('500', 'currency');
      expect(customRef.current.state.isValid).toBe(true);
    });

    fireEvent.changeText(input, '10');
    fireEvent(input, 'blur', {
      target: {
        value: null,
      },
    });

    await waitFor(() => {
      expect(onChangeTextMock).toHaveBeenCalledWith('10', 'currency');
      expect(customRef.current.state.isValid).toBe(false);
    });
  });

  test('should isValid false when input number is above maxvalue', async () => {
    const updateStateMock = jest.spyOn(WmCurrency.prototype, 'updateState');
    const customRef = createRef();
    const tree = render(
      <WmCurrency
        {...defaultProps}
        maxvalue={1000}
        ref={customRef}
        updateon="default"
      />
    );
    const input = tree.getByPlaceholderText('Enter amount');

    fireEvent.changeText(input, '500');
    fireEvent(input, 'blur', {
      target: {
        value: null,
      },
    });

    await waitFor(() => {
      expect(updateStateMock).toHaveBeenCalledWith({ isValid: true });
      expect(customRef.current.state.isValid).toBe(true);
    });

    fireEvent.changeText(input, '50011');
    fireEvent(input, 'blur', {
      target: {
        value: null,
      },
    });

    await waitFor(() => {
      expect(updateStateMock).toHaveBeenCalledWith({ isValid: false });
      expect(customRef.current.state.isValid).toBe(false);
    });
  });
});
