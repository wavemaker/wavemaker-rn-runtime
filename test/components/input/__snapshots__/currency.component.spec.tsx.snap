// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Currency component should have default value for "platform=web" 1`] = `
<View
  style={
    {
      "backgroundColor": "#ffffff",
      "borderColor": "#dddddd",
      "borderRadius": 6,
      "borderStyle": "solid",
      "borderWidth": 1,
      "flexDirection": "row",
      "minHeight": 42,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#4263eb",
        "borderBottomLeftRadius": 6,
        "borderTopLeftRadius": 6,
        "display": "flex",
        "justifyContent": "center",
        "minHeight": 42,
        "paddingBottom": 8,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 8,
        "width": 48,
      }
    }
  >
    <Text
      style={
        {
          "color": "#ffffff",
          "fontFamily": "Roboto",
          "fontSize": 16,
          "fontWeight": "bold",
          "marginVertical": 0.7,
          "textAlign": "center",
          "textAlignVertical": "center",
        }
      }
    >
      $
    </Text>
  </View>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <TextInput
      accessibilityLabel="undefined_i"
      activeFloatingLabelStyle={{}}
      allowContentSelection={true}
      background={
        <React.Fragment>
          <BackgroundComponent
            style={
              {
                "borderRadius": 6,
              }
            }
          />
        </React.Fragment>
      }
      caretHidden={false}
      contextMenuHidden={false}
      editable={true}
      floatingLabelStyle={{}}
      isInputFocused={false}
      keyboardType="numeric"
      onBlur={[Function]}
      onChange={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      onKeyPress={[Function]}
      onSelectionChange={[Function]}
      placeholder="Enter amount"
      placeholderTextColor="#bbbbbb"
      selection={null}
      style={
        [
          [
            {
              "borderColor": "transparent",
              "borderStyle": "solid",
              "borderWidth": 1,
              "flex": 1,
              "minHeight": 42,
              "paddingBottom": 8,
              "paddingLeft": 16,
              "paddingRight": 16,
              "paddingTop": 8,
            },
            {
              "fontFamily": "Roboto",
              "fontSize": 16,
              "userSelect": "text",
            },
            {},
            {},
          ],
          {},
        ]
      }
      testID="undefined_i"
      value="sample text"
    />
  </View>
</View>
`;

exports[`Test Currency component should have default value for native platform 1`] = `
<View
  style={
    {
      "backgroundColor": "#ffffff",
      "borderColor": "#dddddd",
      "borderRadius": 6,
      "borderStyle": "solid",
      "borderWidth": 1,
      "flexDirection": "row",
      "minHeight": 42,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#4263eb",
        "borderBottomLeftRadius": 6,
        "borderTopLeftRadius": 6,
        "display": "flex",
        "justifyContent": "center",
        "minHeight": 42,
        "paddingBottom": 8,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 8,
        "width": 48,
      }
    }
  >
    <Text
      style={
        {
          "color": "#ffffff",
          "fontFamily": "Roboto",
          "fontSize": 16,
          "fontWeight": "bold",
          "marginVertical": 0.7,
          "textAlign": "center",
          "textAlignVertical": "center",
        }
      }
    >
      $
    </Text>
  </View>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <TextInput
      accessibilityValue={
        {
          "max": null,
          "min": null,
        }
      }
      accessible={true}
      activeFloatingLabelStyle={{}}
      allowContentSelection={true}
      background={
        <React.Fragment>
          <BackgroundComponent
            style={
              {
                "borderRadius": 6,
              }
            }
          />
        </React.Fragment>
      }
      caretHidden={false}
      contextMenuHidden={false}
      defaultValue="sample text"
      editable={true}
      floatingLabelStyle={{}}
      isInputFocused={false}
      keyboardType="numeric"
      onBlur={[Function]}
      onChange={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      onKeyPress={[Function]}
      onSelectionChange={[Function]}
      placeholder="Enter amount"
      placeholderTextColor="#bbbbbb"
      selection={null}
      style={
        [
          [
            {
              "borderColor": "transparent",
              "borderStyle": "solid",
              "borderWidth": 1,
              "flex": 1,
              "minHeight": 42,
              "paddingBottom": 8,
              "paddingLeft": 16,
              "paddingRight": 16,
              "paddingTop": 8,
            },
            {
              "fontFamily": "Roboto",
              "fontSize": 16,
              "userSelect": "text",
            },
            {},
            {},
          ],
          {},
        ]
      }
      testID="undefined_i"
    />
  </View>
</View>
`;

exports[`Test Currency component should render correctly with default props 1`] = `
<View
  style={
    {
      "backgroundColor": "#ffffff",
      "borderColor": "#dddddd",
      "borderRadius": 6,
      "borderStyle": "solid",
      "borderWidth": 1,
      "flexDirection": "row",
      "minHeight": 48,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#4263eb",
        "borderBottomLeftRadius": 6,
        "borderTopLeftRadius": 6,
        "display": "flex",
        "justifyContent": "center",
        "minHeight": 42,
        "paddingBottom": 8,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 8,
        "width": 48,
      }
    }
  >
    <Text
      style={
        {
          "color": "#ffffff",
          "fontFamily": "Roboto",
          "fontSize": 16,
          "fontWeight": "bold",
          "marginVertical": 0.7,
          "textAlign": "center",
          "textAlignVertical": "center",
        }
      }
    >
      $
    </Text>
  </View>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      collapsable={false}
      pointerEvents="none"
      style={
        {
          "color": "#bbbbbb",
          "fontSize": 14,
          "left": 16,
          "position": "absolute",
          "textAlign": "left",
          "top": 12,
          "transform": [
            {
              "translateY": 0,
            },
            {
              "translateX": 0,
            },
            {
              "scale": 1,
            },
          ],
          "width": 160,
          "zIndex": 1,
        }
      }
    >
      <Text
        ellipsizeMode="tail"
        style={
          {
            "color": "#bbbbbb",
            "fontFamily": undefined,
            "fontSize": 14,
            "fontStyle": undefined,
            "fontVariant": undefined,
            "fontWeight": undefined,
            "includeFontPadding": undefined,
            "letterSpacing": undefined,
            "lineHeight": undefined,
            "textAlign": "left",
            "textAlignVertical": undefined,
            "textDecorationColor": undefined,
            "textDecorationStyle": undefined,
            "textShadowColor": undefined,
            "textShadowOffset": undefined,
            "textShadowRadius": undefined,
            "textTransform": undefined,
            "userSelect": undefined,
            "verticalAlign": undefined,
            "writingDirection": undefined,
          }
        }
      >
        Amount
      </Text>
    </View>
    <TextInput
      accessibilityValue={
        {
          "max": null,
          "min": null,
        }
      }
      accessible={true}
      activeFloatingLabelStyle={
        {
          "color": "#4263eb",
        }
      }
      allowContentSelection={true}
      background={
        <React.Fragment>
          <BackgroundComponent
            style={
              {
                "borderRadius": 6,
              }
            }
          />
        </React.Fragment>
      }
      caretHidden={false}
      contextMenuHidden={false}
      defaultValue=""
      editable={true}
      floatingLabel="Amount"
      floatingLabelStyle={
        {
          "color": "#bbbbbb",
          "fontSize": 14,
          "left": 16,
          "position": "absolute",
          "textAlign": "left",
          "top": 12,
        }
      }
      isInputFocused={false}
      keyboardType="numeric"
      onBlur={[Function]}
      onChange={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      onKeyPress={[Function]}
      onSelectionChange={[Function]}
      placeholder=""
      placeholderTextColor="#bbbbbb"
      selection={null}
      style={
        [
          [
            {
              "borderColor": "transparent",
              "borderStyle": "solid",
              "borderWidth": 1,
              "flex": 1,
              "minHeight": 42,
              "paddingBottom": 8,
              "paddingLeft": 16,
              "paddingRight": 16,
              "paddingTop": 8,
            },
            {
              "fontFamily": "Roboto",
              "fontSize": 16,
              "userSelect": "text",
            },
            {},
            {},
          ],
          {},
        ]
      }
      testID="undefined_i"
    />
  </View>
</View>
`;
